service/header/core_listener.go

```diff
@@ -38,93 +38,59 @@ func NewCoreListener(p2pSub *P2PSubscriber, fetcher *core.BlockFetcher, dag form
 	cl.cancel()
-	cl.ctx, cl.cancel = nil, nil
-	return nil
+	cl.blockSub = nil
+	return cl.fetcher.UnsubscribeNewBlockEvent(ctx)
 }
 
 // listen kicks off a loop, listening for new block events from Core,
 // generating ExtendedHeaders and broadcasting them to the header-sub
 // gossipsub network.
-func (cl *CoreListener) listen(ctx context.Context) {
-	// start subscription to core node new block events
-	err := cl.startBlockSubscription(ctx)
-	if err != nil {
-		log.Errorw("core-listener: starting subscription to core client", "err", err)
-		return
-	}
-
+func (cl *CoreListener) listen() {
 	for {
 		select {
-		case <-cl.ctx.Done():
-			if err := cl.cancelBlockSubscription(context.Background()); err != nil {
-				log.Errorw("core-listener: canceling subscription to core", "err", err)
+		case b, ok := <-cl.blockSub:
+			if !ok {
+				return
 			}
-			return
-		default:
-			// get next header from core
-			next, err := cl.nextHeader(cl.ctx)
+
+			comm, vals, err := cl.fetcher.GetBlockInfo(cl.ctx, &b.Height)
 			if err != nil {
-				log.Errorw("core-listener: getting next header", "err", err)
-				return
+				log.Errorw("core-listener: getting block info", "err", err)
+				continue
 			}
-			if next == nil {
-				log.Debugw("core-listener: no subsequent header, exiting listening service")
-				return
+
+			eh, err := MakeExtendedHeader(cl.ctx, b, comm, vals, cl.dag)
+			if err != nil {
+				log.Errorw("core-listener: making extended header", "err", err)
+				continue
 			}
+
 			// broadcast new ExtendedHeader
-			err = cl.p2pSub.Broadcast(cl.ctx, next)
+			err = cl.p2pSub.Broadcast(cl.ctx, eh)
 			if err != nil {
-				log.Errorw("core-listener: broadcasting next header", "height", next.Height,
+				log.Errorw("core-listener: broadcasting next header", "height", eh.Height,
 					"err", err)
-				return
 			}
+		case <-cl.ctx.Done():
+			return
 		}
 	}
 }
-
-// startBlockSubscription starts the CoreListener's subscription to new block events
-// from Core.
-func (cl *CoreListener) startBlockSubscription(ctx context.Context) error {
-	if cl.blockSub != nil {
-		return fmt.Errorf("subscription already started")
-	}
-
-	var err error
-	cl.blockSub, err = cl.fetcher.SubscribeNewBlockEvent(ctx)
-
-	return err
-}
-
-// cancelBlockSubscription stops the CoreListener's subscription to new block events
-// from Core.
-func (cl *CoreListener) cancelBlockSubscription(ctx context.Context) error {
-	return cl.fetcher.UnsubscribeNewBlockEvent(ctx)
-}
-
-// nextHeader returns the next latest header from Core.
-func (cl *CoreListener) nextHeader(ctx context.Context) (*ExtendedHeader, error) {
-	select {
-	case newBlock, ok := <-cl.blockSub:
-		if !ok {
-			return nil, fmt.Errorf("subscription closed")
-		}
-
-		comm, vals, err := cl.fetcher.GetBlockInfo(ctx, &newBlock.Height)
-		if err != nil {
-			return nil, err
-		}
-
-		return MakeExtendedHeader(ctx, newBlock, comm, vals, cl.dag)
-	case <-ctx.Done():
-		return nil, ctx.Err()
-	}
```

Comment:
Each select adds not just cpu overhead, but also cognitive. This is the case of totally redundant select being extracted separately. For some reason, you also check the same `ctx.Done` in here and in the `listen`. Again, harder times to review and harder times CPU,